### PR TITLE
[material_ui] Fix carousel view builder arguments

### LINKED ISSUE #188575: [Bug Report] `CarouselView.weightedBuilder` does not respect supplied arguments
### Steps to reproduce

- Use `CarouselView.weightedBuilder` in the app.
- Notice that some arguments are not used at all.

Please refer to the attached code sample.

### Expected results

The `CarouselView.weightedBuilder` constructor should produce same visual output as `CarouselView.weighted` (except using the item builder in code). Following arguments should work as expected:

- `padding`
- `backgroundColor`
- `elevation`
- `shape`
- `itemClipBehavior`
- `overlayColor`
- `onTap`
- `enableSplash`

### Actual results

None of the arguments listed above work.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(home: Repro()));
}

class Repro extends StatelessWidget {
  const Repro({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SizedBox(
          height: 300,
          child: CarouselView.weightedBuilder(
            flexWeights: const <int>[1, 7, 1],
            itemCount: 10,
            padding: const EdgeInsets.all(12),
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(32),
            ),
            itemClipBehavior: Clip.antiAlias,
            backgroundColor: Colors.red,
            elevation: 8,
            overlayColor: WidgetStateProperty.all(
              Colors.black.withOpacity(0.2),
            ),
            onTap: (int index) {
              debugPrint('Tapped $index');
            },
            itemBuilder: (BuildContext context, int index) {
              return Image.network(
                'https://picsum.photos/seed/$index/600/600',
                fit: BoxFit.cover,
              );
            },
          ),
        ),
      ),
    );
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/user-attachments/assets/513b34cc-1ece-4b1c-879d-080d80041f63

</details>

### Logs

N/A

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.44.3, on macOS 26.5.1 25F80 darwin-arm64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 35.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 26.5)
[✓] Chrome - develop for the web
[✓] Connected device (3 available)
[✓] Network resources

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by Mairramer**
Fixes https://github.com/flutter/flutter/issues/188575

Port of https://github.com/flutter/flutter/pull/188578

## Pre-Review Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [AI contribution guidelines] and understand my responsibilities, or I am not using AI tools.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter].
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I followed [the version and CHANGELOG instructions], using [semantic versioning] and the [repository CHANGELOG style], or I have commented below to indicate which documented exception this PR falls under[^1].
- [x] I updated/added any relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or I have commented below to indicate which [test exemption] this PR falls under[^1].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

[^1]: Regular contributors who have demonstrated familiarity with the repository guidelines only need to comment if the PR is not auto-exempted by repo tooling.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[AI contribution guidelines]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#ai-contribution-guidelines
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[the auto-formatter]: https://github.com/flutter/packages/blob/main/script/tool/README.md#format-code
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[the version and CHANGELOG instructions]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version-and-changelog-updates
[semantic versioning]: https://dart.dev/tools/pub/versioning#semantic-versions
[repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[test exemption]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/material_ui/lib/src/carousel.dart [hunk 1]
```diff
@@ -687,34 +681,41 @@ class _CarouselViewState extends State<CarouselView> {
...
       if (widget.infinite && widget.itemCount != null && widget.itemCount! > 0) {
-        final int itemCount = widget.itemCount!;
-        effectiveBuilder = (BuildContext context, int index) {
-          return widget.itemBuilder!(context, index % itemCount);
-        };
-      } else {
-        effectiveBuilder = widget.itemBuilder!;
+        itemIndex = index % widget.itemCount!;
       }
+      child = widget.itemBuilder!(context, itemIndex);
```
[the remark this answers is not in the document]
Comment (TDuffinNTU) [reply packages/material_ui/lib/src/carousel.dart]: >To prevent potential out-of-bounds or null-pointer errors in the builder,

Just FYI this is existing behaviour. I'd argue this is okay to leave as-is, since the developer can just return `null` from `widget.itemBuilder` if the index is invalid - which is the existing behaviour.

Also, It may be that you'd want to handle an invalid index especially for a carousel widget rather than simply omit it, but that's my personal opinion.

TL;DR The suggested change miiiight be consdered "breaking" even if the suggestion is technically correct.

[the remark this answers is not in the document]
Comment (TDuffinNTU) [reply packages/material_ui/lib/src/carousel.dart]: In fact, this behaviour is documented:

```dart
  /// Called to build carousel item on demand.
  ///
  /// Will be called only for indices greater than or equal to zero and less
  /// than [itemCount] (if [itemCount] is non-null).
  ///
  /// Should return null if asked to build a widget with a greater index than
  /// exists.
  final NullableIndexedWidgetBuilder? itemBuilder;
```

Maybe updating the comment would be useful ie:
>  // Should return null if asked to build a widget with an **out-of-range or otherwise-invalid index**

But, again, clearly the intention of the existing API is to **tell developers to consider invalid indices**, rather than detect them within the widget itself.

File: packages/material_ui/lib/src/carousel.dart [hunk 2]
```diff
@@ -687,34 +681,45 @@ class _CarouselViewState extends State<CarouselView> {
...
+        child = null;
+      } else {
+        if (widget.infinite) {
+          itemIndex = index % widget.children.length;
+        }
+        child = widget.children[itemIndex];
+      }
     }
 
+    return child != null ? _buildCarouselItem(child, itemIndex) : null;
```
Comment (QuncCccccc) [inline packages/material_ui/lib/src/carousel.dart:708]: Could we add an infinite-carousel test for this wrapped index? For example, use three items, scroll to a repeated copy of item 0, tap it, and verify that onTap receives 0 rather than the unbounded sliver index. The current tests only tap the initial item, so they would still pass if the raw index were passed here. It would be useful to cover both `CarouselView.builder` and `CarouselView.weightedBuilder`.

File: packages/material_ui/lib/src/carousel.dart [hunk 3]
```diff
@@ -687,34 +681,45 @@ class _CarouselViewState extends State<CarouselView> {
...
+    Widget? child;
 
-    NullableIndexedWidgetBuilder effectiveBuilder;
     if (widget.itemBuilder != null) {
-      if (widget.infinite && widget.itemCount != null && widget.itemCount! > 0) {
-        final int itemCount = widget.itemCount!;
-        effectiveBuilder = (BuildContext context, int index) {
-          return widget.itemBuilder!(context, index % itemCount);
-        };
+      if (widget.itemCount == 0) {
```
Comment (QuncCccccc) [inline packages/material_ui/lib/src/carousel.dart:689]: Is `itemCount: 0` an intended supported case here? If so, could we add a test for it and ensure controller methods such as `animateToItem` also handle an empty carousel?

File: packages/material_ui/lib/src/carousel.dart [hunk 4]
```diff
@@ -687,34 +681,45 @@ class _CarouselViewState extends State<CarouselView> {
...
+        }
+        child = widget.children[itemIndex];
+      }
     }
 
+    return child != null ? _buildCarouselItem(child, itemIndex) : null;
+  }
+
+  Widget _buildSliverCarousel(ThemeData theme) {
+    // Determine the child count and builder based on whether we're using lazy loading
```
Comment (QuncCccccc) [inline packages/material_ui/lib/src/carousel.dart]: Maybe
```suggestion
    // // Determine the child count based on whether an item builder is used.
```

File: packages/material_ui/example/test/carousel/carousel.1_test.dart [hunk 5]
```diff
@@ -13,23 +13,54 @@ void main() {
...
 
     expect(find.byType(CarouselView), findsOneWidget);
-
     expect(find.text('Item 0'), findsOneWidget);
 
     expect(find.text('Item 999'), findsNothing);
 
     final Finder carousel = find.byType(CarouselView);
-    await tester.drag(carousel, const Offset(-350, 0));
+    await tester.drag(carousel, const Offset(-400, 0));
```
Comment (QuncCccccc) [inline packages/material_ui/example/test/carousel/carousel.1_test.dart:25]: Is this drag-distance change related to the fix? Since itemExtent is still 350, it would be helpful to explain why the test now needs to drag 400 pixels.

**Review by QuncCccccc — COMMENTED** [review]
Thanks for the fix! I left some comments below.

File: packages/material_ui/lib/src/carousel.dart (hunk 3 above)
Comment (Mairramer) [reply packages/material_ui/lib/src/carousel.dart:689]: Yes, `itemCount: 0` (or `children.isEmpty`) is intended to be supported.

File: packages/material_ui/example/test/carousel/carousel.1_test.dart (hunk 5 above)
Comment (Mairramer) [reply packages/material_ui/example/test/carousel/carousel.1_test.dart:25]: Yes, exactly. With this fix, the flow now correctly goes through `_buildCarouselItem`, so the item is properly wrapped with `Padding`, `Material`, and the other expected widgets.
